Crash reporter

white-axe opened the pull request:
**Connections**
* Closes #6

**Description**
Luminol now restarts after panicking on both native and web platforms. After restarting, it shows a window asking the user if they want to send the crash report to the developers. The crash report is a JSON object sent as a POST request; it contains the panic message as well as the Git revision, target triple and profile (debug/release).

It reports to http://localhost:3246 right now for easy testing. This should be changed in crates/ui/src/windows/reporter.rs to a real URL at some point. Don't worry about CORS - web builds can send crash reports even if the endpoint doesn't set CORS headers.

Restarting after panics is done on Unix-like systems by using one of the exec system calls to replace the current process without creating a new one. On Windows, it spawns a background process and then terminates the original one. For WebAssembly, it's done by reloading the WebAssembly modules. Reports are persisted across restarts in native builds by creating a temporary file and in web builds by saving the report as a JavaScript global variable.

I was originally going to try using `std::panic::catch_unwind` to catch panics, but this doesn't work in WebAssembly due to the panic_abort panic handler and it causes segfaults and other signs of undefined behaviour in native builds somewhere inside of winit.

There are also some related improvements to web builds:

- Luminol now completely caches itself after loading for the first time so that the browser doesn't have to download the WebAssembly module again every time a panic occurs. No internet access is required after the first page load.
- The Git revision is now determined from the Trunk hooks instead of with the git-version macro so that it can correctly determine if the code has uncommitted changes. It should now append or not append "-modified" to the Git revision properly instead of always appending "-modified" regardless of whether there are uncommitted changes.
- Event handlers for mouse and keyboard and the like are now unregistered after panics for performance reasons.
- Clippy no longer errors out when checking luminol-web, so we can now see clippy lints for crates that depend on luminol-web.

**Testing**
Right now, a panic can be triggered by attempting to create or edit events in the map editor.

Testing for reception of reports on localhost can be done with a simple echo server like this one: https://github.com/cdfuller/echo-server. If you use this one in particular, make sure to run it with the `-v` flag so that it prints the contents of requests. Whatever echo server is used, test that it is able to receive reports from both native and web builds of Luminol.

It would also be a good idea to check that a real, internet-facing server is able to receive reports, especially that one that does not support CORS is able to receive reports from a web build of Luminol.

**Checklist**

- [x] Run `cargo fmt`.
- [x] Run `cargo clippy`. If applicable, add:
  - [x] `--target wasm32-unknown-unknown`
- [x] Run `cargo build --release` 
- [x] If applicable, run `trunk build --release`

white-axe committed:
feat(native): restart program after panicking on Unix-like systems

white-axe committed:
fix(native): use `std::env::current_exe()` to get executable path

The program's first argument is not guaranteed to be a path to the
current executable.

Also, at least on Linux, `std::env::current_exe()` is able to correctly
find the new location of the executable if the user moves the executable
before the panic handler runs.

white-axe committed:
feat(native): add fallback restart code for non-Unix systems

white-axe committed:
feat(native): persist panic reports by saving to temporary files

white-axe committed:
perf(web): panic channel now uses oneshot instead of flume

white-axe committed:
feat(web): restart program and persist panic report after panicking

white-axe committed:
perf(web): terminate worker thread after panicking

white-axe committed:
merge: merge branch 'dev' into reporter

white-axe committed:
perf(web): unregister event listeners on panic

white-axe committed:
chore(web): remove extra use in web_runner.rs to reduce conflicts

white-axe committed:
refactor(web): remove unnecessary static variable from main.rs

white-axe committed:
fix(web): set `clippy::missing_safety_doc` to `warn`

Setting this to `forbid` seems to break clippy when using wasm-bindgen.
I'm setting this to `warn` so we can see clippy lints for crates that
depend on luminol-web in web builds.

white-axe committed:
chore(web): fix clippy warnings

white-axe committed:
feat: implement most of the reporter UI

white-axe committed:
chore: rustfmt

white-axe committed:
feat: implement sending reports

white-axe committed:
chore(web): update coi-serviceworker to commit 7b1d2a092d0d2dd2b7270b6f12f13605de26f214

white-axe committed:
fix(web): reporter now uses no-cors when sending reports

white-axe committed:
fix: fix reporter window height being too small

white-axe committed:
fix(native): always print full backtraces

white-axe committed:
fix: don't restart after panicking if UI failed to start

white-axe committed:
fix(web): compute Git version before running Trunk hooks

Without this, the Git version will always have "-modified" at the end in
web builds even if the code is unmodified, because the Trunk hooks
modify .cargo/config.toml.

white-axe committed:
chore: rename `luminol_version` to `luminol_revision`

white-axe committed:
feat(web): service worker now caches Luminol's files

Luminol can now be run offline after loading it in a browser once.

white-axe committed:
chore(web): clippy

white-axe committed:
style: use declarative `cfg` macro for report `debug` field

white-axe committed:
fix(web): fix environment being inconsistent between page reloads

This fixes a problem where sometimes, either cross-origin isolation
would not be enabled at all or sw.js would have its own COEP (but not
the COEP for other files) set to require-corp instead of credentialless.
This would be fixed by refreshing the page a number of times.

Now, COI should always be enabled and the COEP for sw.js should always
be credentialless, without the user having to refresh the page: the page
is automatically refreshed.

white-axe committed:
chore: rustfmt

white-axe committed:
fix(native): remove report file after failing to restart

white-axe committed:
fix: don't persist reporter's scroll position

white-axe committed:
fix(web): disable credentialless COEP

Haha, turns out we don't even need credentialless COEP to send no-cors
requests! The require-corp COEP works just fine.

white-axe committed:
fix(web): undo last commit

After further testing, it seems Firefox does require credentialless COEP
for this type of no-cors request, so Chromium allowing it is probably a
bug.

white-axe committed:
fix(web): increase robustness of the service worker

This fixes some edge cases I experienced when testing in Firefox where
the user refreshing the page sometimes causes the page's COEP to revert
to "require-corp" (the service worker is supposed to set it to
"credentialless").

I hate web development so much.

white-axe committed:
fix(web): fix Git revision not always being updated

This fixes a bug where updating one of the HTML or JavaScript files
(which changes the output of the `luminol_core::version!` macro) doesn't
cause every crate that uses the `luminol_core::version!` macro to be
recompiled, resulting in parts of Luminol getting different Git revision
strings from this macro.

somedevfox commented on crates/core/Cargo.toml:
Changing git-version from pulling a version from the workspace to a specific version can result in incompatibility between the core and other crates in the future.

white-axe replied:
I didn't change how the git version is determined in native builds. In web builds, I only changed it to determine the git version inside of hooks/trunk_enable_build_std_pre.sh because the git-version macro incorrectly says that the code has been modified relative to the latest commit even when it hasn't been. So the version is still automatically determined.

white-axe committed:
merge: merge branch 'dev' into reporter

white-axe committed:
merge: merge branch 'dev' into reporter

white-axe committed:
refactor: move Git version into `UpdateState`

This prevents the Git revision from being miscalculated if some of the
crates are at a different version than others.

white-axe commented:
Since I'd already changed how the Git version is calculated in web builds, I've also moved the Git version into the luminol crate (#89). This will prevent the version from being different in different crates if the crates are not all at the same revision.

white-axe committed:
style: remove unnecessary `thread_local` in `luminol_core`

white-axe committed:
merge: merge branch 'dev' into reporter

white-axe committed:
merge: merge branch 'dev' into reporter

white-axe committed:
feat: add wgpu backend to report info

Since wgpu errors result in panics, this seems like a useful thing to
have.

white-axe committed:
merge: merge branch 'dev' into reporter

white-axe committed:
chore: fix an update-induced compiler error

white-axe committed:
style: use `UiExt` in reporter window

white-axe commented:
I can't merge this until @somedevfox approves
